libsignal-protocol/src/raw_ptr.rs

```diff
@@ -1,62 +1,97 @@
+use std::ptr::NonNull;
+
 /// A wrapper around `libsignal-protocol`'s native reference counted pointers.
 #[derive(Debug)]
-pub struct Raw<T: SignalType>(*mut T);
+pub struct Raw<T: IsA<sys::signal_type_base>>(NonNull<T>);
 
-impl<T: SignalType> Raw<T> {
-    /// Create a new [`Raw<T>`] from an owned pointer (doesn't affect reference
+impl<T: IsA<sys::signal_type_base>> Raw<T> {
+    /// Create a new [`Raw`] from an owned pointer (doesn't affect reference
     /// count).
-    pub fn from_ptr(raw: *mut T) -> Raw<T> { Raw(raw) }
-
-    /// Create a new [`Raw<T>`] after bumping the reference count.
-    pub fn copied_from(raw: *mut T) -> Raw<T> {
-        unsafe {
-            sys::signal_type_ref(SignalType::as_signal_base(raw));
-        }
+    ///
+    /// # Safety
+    ///
+    /// This assumes `raw` points to a valid, non-null instance of `T`.
+    pub unsafe fn from_ptr(raw: *mut T) -> Raw<T> {
+        debug_assert!(!raw.is_null());
+        Raw(NonNull::new_unchecked(raw))
+    }
 
+    /// Create a new [`Raw`] after bumping the reference count.
+    ///
+    /// # Safety
+    ///
+    /// This assumes `raw` points to a valid, non-null instance of `T`.
+    pub unsafe fn copied_from(raw: *mut T) -> Raw<T> {
+        sys::signal_type_ref(T::upcast(raw));
         Raw::from_ptr(raw)
     }
 
-    pub fn as_ptr(&self) -> *mut T { self.0 }
+    /// Go from a `*mut Child` to a `*mut Parent`.
+    ///
+    /// # Safety
+    ///
+    /// This call relies on the [`IsA`] implementation being sound.
+    pub fn upcast<Parent>(self) -> Raw<Parent>
+    where
+        T: IsA<Parent>,
+        Parent: IsA<sys::signal_type_base>,
+    {
+        // safety: self.0 is guaranteed to point to a valid `T` instance.
+        // It also assums the `upcast()` method is sound.
+        unsafe { Raw::from_ptr(T::upcast(self.into_raw())) }
+    }
+
+    pub fn as_ptr(&self) -> *mut T { self.0.as_ptr() }
+
+    /// Consume this [`Raw`] and extract the underlying pointer **without**
+    /// decrementing its reference count.
+    pub fn into_raw(self) -> *mut T {
```

Comment:
Hahahaha, it's funny because the name of the function is the same of the struct 
😂 

Maybe `into_inner` or `unwrap` would be better in that specific case?

> I know that it's convenient for Rust structs to have `into_raw` methods to getting the underlying data structure.

Reply:
`into_inner` sounds like a good name for it! I usually associate `unwrap` with panicking if a particular constraint isn't upheld.